Make docs clearer and improve private repo code (#10)
sterlingwes/decap-proxy: README.md

A Cloudflare Worker Github OAuth proxy for [Decap CMS](https://github.com/decaporg/decap-cms). This allows for deploying Decap without the Netlify Identity or Git Gateway services required to handle Github authentication so that the CMS client can make Github API calls.

## Overview

Suppose you have a website that is generated from some source files on GitHub and you want to provide an admin page where people who lack an understanding of git or HTML can edit the source files using a friendly browser-based interface. Decap CMS provides such an interface and can be configured to work with Github as the back-end, using Cloudflare to serve an [OAuth Proxy](https://decapcms.org/docs/backends-overview/#using-github-with-an-oauth-proxy) that will act as an authentication go-between.

> This repository provides the elements needed to deploy such an OAuth Proxy on a Cloudflare Worker node. The proxy will be deployed to its own subdomain, separate from whatever website domain you're using Decap with. That subdomain, which will be used to form the **PROXY URL**, below, will be used exclusively for hosting the OAuth Proxy. Either you will be using the `workers.dev` domain that Cloudflare automatically gives your workers, or you will be providing a custom subdomain (eg. `proxy.mydomain.com`) for a domain zone you own.

Reviewer: ```suggestion
This repository provides the elements needed to deploy such an OAuth Proxy on a Cloudflare Worker node. The proxy will be deployed to its own subdomain, separate from whatever website domain you're using Decap to manage. That subdomain will be referenced by the **PROXY URL**, below, and used exclusively for hosting the OAuth Proxy. Either you will be using the `workers.dev` domain that Cloudflare automatically gives your workers, or you will be providing a custom subdomain (eg. `proxy.mydomain.com`) for a domain zone you own.
```

> ```mermaid

Reviewer: 😍 love this!

graph TD;

subgraph GitHub
    R[(Repo)];
    A[/OAuth App\];
end

P["OAuth Proxy<br>(Cloudflare worker, eg at decap.mydomain.com)"]

subgraph Website
	X@{ shape: braces, label: "Could be hosted<br>on Cloudflare Pages<br>eg at www.mydomain.com" };
	D@{ shape: doc, label: "DecapCMS Panel<br>(maybe at /admin)" }
    Y@{ shape: docs, label: "Other pages" }
end

A-- callback -->P;
P-->A;
D-- auth -->P;
R-->Y;

%% Omitting the following link that results in ugly layout
%% D<-->R;
```

> Decap CMS  assumes that you have some sort of CI/CD pipleine set up so that when commits are made to your github source, your website will be rebuilt, automatically. If you want to use a Cloudflare-based solution for that side of things, you could consider using Cloudflare Pages, or even Cloudflare Workers. Any such setups are not covered by this document, but it's recommended that you have that working before proceeding.

Reviewer: ```suggestion
Decap CMS  assumes that you have some sort of CI/CD pipeline set up so that when commits are made to your github source, your website will be rebuilt, automatically. If you want to use a Cloudflare-based solution for that side of things, you could consider using Cloudflare Pages, or even Cloudflare Workers. Any such setups are not covered by this document, but it's recommended that you have that working before proceeding.
```

## 1. Create a Github OAuth App

Once you're logged into the GitHub account that hosts the source to your web-page, you can [configure a Github OAuth application](https://github.com/settings/applications/new). 

* For `Homepage URL`, please enter the **PROXY URL**.
* For `Application callback URL`, please enter the **PROXY URL** followed by `/callback`. For the example in the table above, it would mean using: `https://decap.mydomain.com/callback`.
* The `Application Name` and `Application Description` can be whatever you want them to be. They will not affect the functionality of the GitHub OAuth App.

Save the OAuth client ID and secret - you'll need them in [2.4 Configure the Cloudflare Worker](#24-configure-the-cloudflare-worker), below.

In the future, when you want to revisit the OAuth App you have created here, you can find it under `OAuth Apps` on the github [Developer Settings page](https://github.com/settings/developers).

## 2. Deploy the OAuth Proxy

> You will need some sort of working area where you will be cloning this decap-proxy repo, editing its contents, and being the launching pad for the deployment of the OAugh Proxy. This working area typically consists of a terminal on your local computer where you can run node packages using `npx` and you have a browser available (though the latter is optional).

Reviewer: ```suggestion
You will need some sort of working area where you will be cloning this decap-proxy repo, editing its contents, and performing the deployment of the OAuth Proxy. This working area typically consists of a terminal on your local computer where you can run node packages using `npx` and you have a browser available (though the latter is optional).
```

Here's an overview of the steps needed to deploy the OAuth Proxy. 

1. [Clone decap-proxy](#21-clone-decap-proxy)
2. [Prepare the wrangler.toml file](#22-prepare-the-wranglertoml-file)
3. [Establish Connection to Cloudflare](#23-establish-connection-to-cloudflare)
4. [Configure the Cloudflare Worker](#24-configure-the-cloudflare-worker)
5. [Deploy the Proxy on the Worker](#25-deploy-the-proxy-on-the-worker)

Details follow.

### 2.1 Clone decap-proxy

Use whatever git tools you would normally use to clone this repository to your working area. Most people would be able to use the following command in their terminal: `git clone https://github.com/sterlingwes/decap-proxy`.

### 2.2 Prepare the wrangler.toml file

> The first step is to copy wrangler.toml.sample to wrangler.toml. For many people, this would mean running the following command in their terminal: `cp wrangler.toml.sample wrangler.toml`. You can look at the commented out directives in the resulting file and see if there are any unusual configuration of Cloudflare you wish to enact. But for most people, especially as they are getting started, you only need to consider the following four changes.

Reviewer: ```suggestion
The first step is to copy wrangler.toml.sample to wrangler.toml. For many people, this would mean running the following command in their terminal: `cp wrangler.toml.sample wrangler.toml`. You can look at the commented-out directives in the resulting file and see if there are any unusual configuration of Cloudflare you wish to enact. But for most people, especially when getting started, you only need to consider the following four changes.
```

The *name* parameter at the top defines the name that your worker will be given in your Cloudflare dashboard, and will determine the first part of the default URL for the worker, if you don't disable that. You should change the name to something sensible for you.

If you're using a custom domain in your **PROXY URL**, then alter wrangler.toml by uncommenting and appropriately altering the route directive at the top of the file to use **PROXY DOMAIN** as the `pattern`, and **PROXY ZONE** as the `zone_name`:

```typescript
route = { pattern = "decap.mydomain.com", zone_name = "mydomain.com", custom_domain = true }
```
Make sure you change the `pattern` and `zone_name` to be consistent with your **PROXY URL**.

Even with that change, the OAuth proxy will be available at the default workers.dev domains that are available for workers. You can disable that by uncommenting this line at the top of wrangler.toml:

```typescript
workers_dev = false
```

> Finally, if the github repository that holds the source to your web-site is a private repository, then you should change the declaration of the `GITHUB_REPO_PRIVATE` in wrangler.toml to set the value to 1.

Reviewer: ```suggestion
Finally, if the github repository that holds the source to your website is a private repository, then you should change the declaration of the `GITHUB_REPO_PRIVATE` in wrangler.toml to set the value to 1.
```
